docs/proposals/diagnose-experience.md

```diff
@@ -0,0 +1,250 @@
+  - Example: `DEBUG: Attempting to authenticate user [UserID: usr123]`  (exclude authentication token and password information).
+
+## Proposals for ORAS CLI
+
+Based on the concepts and conventions above, here are the proposal for ORAS diagnose experience improvement: 
+
+- Deprecate the `--verbose` flag and keep `--debug` flag to avoid ambiguity. It is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs as it is in ORAS. Meanwhile, This change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.
+- Make the verbose output of commands `pull`, `push`, `attach` as the default (status) output. See examples at the bottom.
+- Make the verbose output of command  `discover` as a formatted output, controlled by `--format tree-full`. 
+- Add two empty lines as the separator between each request and response for readability.
+- Add timestamp of each request and response to the beginning of each request and response.
+- Upon success, print out the response body in the debug logs if the `mediaType` of the processed OCI object (e.g. image manifest) is JSON or plain text format. The [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) should be included in the response body by default for diagnose purposes.
+- Summarize common conventions for writing clear and analyzable debug logs.
+- Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, except for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on terminal but keeping plain text in a non-TTY mode respectively.
+
+## Investigation on other CLI tools
+
+To make sure the ORAS diagnose functions are natural and easier to use, it worth knowing how diagnose functions work in other popular client tools. 
+
+#### Curl
+
+Curl only has a `--verbose` option to output verbose logs. No `--debug` option.
+
+#### Docker and Podman
+
+Docker provides two options `--debug` and `--log-level`  to control debug logs output within different log levels, such as INFO, DEBUG, WARN, etc. No `--verbose` option. Docker has its own daemon service running in local so its logs might be much more complex.
+
+#### Helm
+
+Helm CLI tool provides a global flag `--debug` to enable verbose output.
+
+## Examples in ORAS
+
+This section lists the current behaviors of ORAS debug logs, proposes the suggested changes to ORAS CLI commands. 
+
+### oras copy
+
+Pick the first two requests and responses as examples:
+
+```
+oras copy ghcr.io/oras-project/oras:v1.2.0 --to-oci-layout oras-dev:v1.2.0 --debug
+```
+
+**Current debug log**
```

Comment:
Should we call out that this logs are for TTY and mention the current logs in non-TTY mode as well?

Reply:
@Wwwsylvia  I mentioned the tiny difference of debug logs between the TTY and non-TTY modes in the Proposals section. Does this description add more clarity? 

> Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, **the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, except for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on terminal but keeping plain text in a non-TTY mode respectively.**

Reply:
I meant we can put examples of the current logs for both TTY and non-TTY.

Reply:
An example of logs for non-TTY can be get from `oras repo ls localhost:5000 --debug 2>&1 | cat`

Reply:
Logs for non-TTY looks like:

> time=2024-10-10T18:45:10+08:00 level=debug msg=Request #0
> > Request URL: "http://localhost:5000/v2/240923test1/manifests/v1"
> > Request method: "HEAD"

Reply:
@Wwwsylvia  The debug logs `--debug` implies non-TTY mode, which is the same as the output with using `--no-tty`. How to print out the debug logs in TTY mode?

Reply:
`oras repo ls localhost:5000 --debug` prints logs in TTY mode

Reply:
Added TTY and non-TTY debug logs per suggestions